src/crimson/os/seastore/onode_manager/staged-fltree/node.cc

```diff
@@ -0,0 +1,805 @@
+      cloned_root->make_root_new(c_other, std::move(super_other));
+    });
+  }).safe_then([this_ref]{});
+}
+
+node_future<Ref<tree_cursor_t>> LeafNode::insert_value(
+    context_t c, const key_hobj_t& key, const onode_t& value,
+    const search_position_t& pos, const MatchHistory& history,
+    match_stat_t mstat) {
+#ifndef NDEBUG
+  if (pos.is_end()) {
+    assert(impl->is_level_tail());
+  }
+#endif
+  logger().debug("OTree::Leaf::Insert: "
+                 "pos({}), {}, {}, {}, mstat({}) ...",
+                 pos, key, value, history, mstat);
+  search_position_t insert_pos = pos;
+  auto [insert_stage, insert_size] = impl->evaluate_insert(
+      key, value, history, mstat, insert_pos);
+  auto free_size = impl->free_size();
+  if (free_size >= insert_size) {
+    // insert
+    on_layout_change();
+    impl->prepare_mutate(c);
+    auto p_value = impl->insert(key, value, insert_pos, insert_stage, insert_size);
+    assert(impl->free_size() == free_size - insert_size);
+    assert(insert_pos <= pos);
+    assert(p_value->size == value.size);
+    auto ret = track_insert(insert_pos, insert_stage, p_value);
+    validate_tracked_cursors();
+    return node_ertr::make_ready_future<Ref<tree_cursor_t>>(ret);
+  }
+  // split and insert
+  Ref<LeafNode> this_ref = this;
+  return (is_root() ? upgrade_root(c) : node_ertr::now()
+  ).safe_then([this, c] {
+    return LeafNode::allocate(c, impl->field_type(), impl->is_level_tail());
+  }).safe_then([this_ref, this, c, &key, &value, &history,
+                insert_pos, insert_stage, insert_size](auto fresh_right) mutable {
+    auto right_node = fresh_right.node;
+    // no need to bump version for right node, as it is fresh
+    on_layout_change();
+    impl->prepare_mutate(c);
+    auto [split_pos, is_insert_left, p_value] = impl->split_insert(
+        fresh_right.mut, *right_node->impl, key, value,
+        insert_pos, insert_stage, insert_size);
+    assert(p_value->size == value.size);
+    track_split(split_pos, right_node);
+    Ref<tree_cursor_t> ret;
+    if (is_insert_left) {
+      ret = track_insert(insert_pos, insert_stage, p_value);
+    } else {
+      ret = right_node->track_insert(insert_pos, insert_stage, p_value);
+    }
+    validate_tracked_cursors();
+    right_node->validate_tracked_cursors();
+
+    // propagate insert to parent
+    return insert_parent(c, right_node).safe_then([ret] {
+      return ret;
+    });
+    // TODO (optimize)
+    // try to acquire space from siblings before split... see btrfs
+  });
+}
+
+node_future<Ref<LeafNode>> LeafNode::allocate_root(
+    context_t c, RootNodeTracker& root_tracker) {
+  return LeafNode::allocate(c, field_type_t::N0, true
+  ).safe_then([c, &root_tracker](auto fresh_node) {
+    auto root = fresh_node.node;
+    return c.nm.get_super(c.t, root_tracker
+    ).safe_then([c, root](auto&& super) {
+      root->make_root_new(c, std::move(super));
+      return root;
+    });
+  });
+}
+
```

Comment:
I'm not sure I get this part.  What if a user wanted two different cursors with independent positions which happen to start at the same position?

Reply:
`Cursor` holds a `Ref<tree_cursor_t>` which can be shared by multiple cursors. And `operator++()` will allocate/point to a different `tree_cursor_t`. See https://github.com/ceph/ceph/blob/55c40e3642572522ecbc4a1ac86f2e49fe946f48/src/crimson/os/seastore/onode_manager/staged-fltree/tree.h#L105

Reply:
Hmm, ok.